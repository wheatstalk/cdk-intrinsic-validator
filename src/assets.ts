import * as path from 'path';

export const LAMBDA_ASSET_DIR = path.join(__dirname, '..', 'lambda');