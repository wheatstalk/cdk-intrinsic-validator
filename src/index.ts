export * from './intrinsic-validator';
export { LAMBDA_ASSET_DIR } from './assets';