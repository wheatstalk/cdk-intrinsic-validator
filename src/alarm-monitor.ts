import * as cdk from 'aws-cdk-lib';
import * as cw from 'aws-cdk-lib/aws-cloudwatch';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as sfn_tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Construct } from 'constructs';
import { LAMBDA_ASSET_DIR } from './assets';

/**
 * A CloudWatch alarm monitor.
 */
export interface IAlarmMonitor {
  /**
   * The Step Functions state machine.
   */
  readonly stateMachine: sfn.IStateMachine;

  /**
   * Create input for a state machine execution.
   */
  createTaskInput(options: AlarmMonitorTaskInputOptions): sfn.TaskInput;
}

/**
 * Task input options.
 */
export interface AlarmMonitorTaskInputOptions {
  /** The alarm to monitor */
  readonly alarm: cw.IAlarm;
  /** The maximum duration to monitor */
  readonly duration: cdk.Duration;
}

/**
 * Provides a State Machine to monitor CloudWatch Alarms.
 */
export class AlarmMonitor extends Construct implements IAlarmMonitor {
  readonly stateMachine: sfn.IStateMachine;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const bugState = new sfn.Fail(this, 'BugState');

    const checkAlarmStatusFunction = new lambda.Function(this, 'CheckAlarmStatusFunction', {
      runtime: lambda.Runtime.NODEJS_14_X,
      handler: 'check-alarm-status.handler',
      code: lambda.Code.fromAsset(LAMBDA_ASSET_DIR),
      logRetention: logs.RetentionDays.ONE_MONTH,
      initialPolicy: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: ['cloudwatch:DescribeAlarms'],
          resources: ['*'],
        }),
      ],
    });

    const checkAlarmStatus = new sfn_tasks.LambdaInvoke(this, 'CheckAlarmStatus', {
      integrationPattern: sfn.IntegrationPattern.REQUEST_RESPONSE,
      lambdaFunction: checkAlarmStatusFunction,
      payload: sfn.TaskInput.fromObject({
        'AlarmName.$': '$.AlarmName',
        'MonitoringDurationSeconds.$': '$.MonitoringDurationSeconds',
        'StartTime.$': '$$.Execution.StartTime',
      }),
      resultPath: '$.CheckAlarmStatus',
      resultSelector: {
        'Status.$': '$.Payload.Status',
      },
    });

    checkAlarmStatus.addCatch(bugState);
    checkAlarmStatus.addRetry({ maxAttempts: 3 });

    const alarmStatusChoice = new sfn.Choice(this, 'AlarmStatusChoice');
    const delayForNextCheckStatus = new sfn.Wait(this, 'DelayThenNextCheckStatus', {
      time: sfn.WaitTime.duration(cdk.Duration.seconds(30)),
    });
    const failBecauseAlarming = new sfn.Fail(this, 'FailBecauseAlarming');
    const noAlarm = new sfn.Succeed(this, 'NoAlarm');

    const definition = sfn.Chain.start(checkAlarmStatus)
      .next(alarmStatusChoice
        .when(
          sfn.Condition.stringEquals('$.CheckAlarmStatus.Status', 'ALARM'),
          failBecauseAlarming,
        )
        .when(
          sfn.Condition.stringEquals('$.CheckAlarmStatus.Status', 'MONITORING'),
          delayForNextCheckStatus
            .next(checkAlarmStatus),
        )
        .when(
          sfn.Condition.stringEquals('$.CheckAlarmStatus.Status', 'NO_ALARM'),
          noAlarm,
        )
        .otherwise(bugState),
      );

    this.stateMachine = new sfn.StateMachine(this, 'StateMachine', {
      definition: definition,
    });
  }

  createTaskInput(options: AlarmMonitorTaskInputOptions): sfn.TaskInput {
    return sfn.TaskInput.fromObject({
      AlarmName: options.alarm.alarmName,
      MonitoringDurationSeconds: options.duration.toSeconds(),
    });
  }
}

/**
 * Creates or re-uses a singleton alarm monitor.
 */
export class SingletonAlarmMonitor extends Construct implements IAlarmMonitor {
  private readonly alarmMonitor: IAlarmMonitor;

  public get stateMachine(): sfn.IStateMachine {
    return this.alarmMonitor.stateMachine;
  }

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const stack = cdk.Stack.of(this);
    this.alarmMonitor = stack.node.tryFindChild('SingletonAlarmMonitor') as AlarmMonitor
      ?? new AlarmMonitor(stack, 'SingletonAlarmMonitor');
  }

  createTaskInput(options: AlarmMonitorTaskInputOptions): sfn.TaskInput {
    return this.alarmMonitor.createTaskInput(options);
  }
}