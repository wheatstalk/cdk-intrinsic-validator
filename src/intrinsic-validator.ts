import * as path from 'path';
import * as cw from '@aws-cdk/aws-cloudwatch';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecs from '@aws-cdk/aws-ecs';
import * as lambda from '@aws-cdk/aws-lambda';
import * as logs from '@aws-cdk/aws-logs';
import * as sfn from '@aws-cdk/aws-stepfunctions';
import * as sfn_tasks from '@aws-cdk/aws-stepfunctions-tasks';
import * as cdk from '@aws-cdk/core';
import * as cr from '@aws-cdk/custom-resources';
import { SingletonAlarmMonitor } from './alarm-monitor';

/**
 * Props for `IntrinsicValidator`
 */
export interface IntrinsicValidatorProps {
  /**
   * Name the Step Functions State Machine.
   * @default - CDK automatically picks a name
   */
  readonly stateMachineName?: string;

  /**
   * Validations to run every time the stack is deployed.
   * @default - no validations are run
   */
  readonly validations?: Validation[];
}

/**
 * Adds intrinsic validation to a CloudFormation stack.
 */
export class IntrinsicValidator extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: IntrinsicValidatorProps = {}) {
    super(scope, id);

    const definition = new sfn.Parallel(this, 'Validations');

    let validationIndex = 0;
    const validations = props.validations ?? [];
    for (const validation of validations) {
      const config = validation._bind(definition, `[${validationIndex}] ${validation._label}`);
      definition.branch(config.chainable);
      validationIndex++;
    }

    if (validations.length === 0) {
      const noValidationsPass = new sfn.Pass(this, 'NoValidations', {
        result: sfn.Result.fromString('Disabled'),
      });
      definition.branch(noValidationsPass);
    }

    const stateMachine = new sfn.StateMachine(this, 'StateMachine', {
      stateMachineName: props.stateMachineName,
      definition,
    });

    const code = lambda.Code.fromAsset(path.join(__dirname, '..', 'lambda'));
    const runtime = lambda.Runtime.NODEJS_14_X;
    const logRetention = logs.RetentionDays.ONE_MONTH;

    const onEventHandler = new lambda.Function(this, 'OnEventHandler', {
      runtime,
      code,
      logRetention,
      handler: 'intrinsic-validator-provider.onEventHandler',
      environment: {
        STATE_MACHINE_ARN: stateMachine.stateMachineArn,
      },
    });
    stateMachine.grantStartExecution(onEventHandler);

    const isCompleteHandler = new lambda.Function(this, 'IsCompleteHandler', {
      runtime,
      code,
      logRetention,
      handler: 'intrinsic-validator-provider.isCompleteHandler',
      environment: {
        STATE_MACHINE_ARN: stateMachine.stateMachineArn,
      },
    });
    stateMachine.grantRead(isCompleteHandler);

    const provider = new cr.Provider(this, 'Provider', {
      onEventHandler,
      isCompleteHandler,
      logRetention,
    });

    // Create a new logical resource every time we deploy. This is how we cause
    // the validation to be run on every deployment.
    const randomness = Math.round(Math.random() * 10000000);
    new cdk.CustomResource(this, `Resource${randomness}`, {
      serviceToken: provider.serviceToken,
    });
  }
}


/** @internal */
export interface ValidationConfig {
  readonly chainable: sfn.IChainable;
}

/**
 * Create many types of validations.
 */
export abstract class Validation {
  /**
   * Create a validation that always succeeds.
   */
  static alwaysSucceeds(): Validation {
    return new AlwaysSucceeds();
  }

  /**
   * Create a validation that always fails.
   */
  static alwaysFails(): Validation {
    return new AlwaysFails();
  }

  /**
   * Create a validation that runs a fargate task and waits for it to succeed.
   */
  static fargateTaskSucceeds(options: FargateTaskSucceedsOptions): Validation {
    return new FargateTaskSucceeds(options);
  }

  /**
   * Create a validation that executes a Step Functions state machine.
   */
  static stateMachineExecutionSucceeds(options: StateMachineExecutionSucceedsOptions): Validation {
    return new StateMachineExecutionSucceeds(options);
  }

  /**
   * Create a validation that monitors an alarm.
   */
  static monitorAlarm(options: MonitorAlarmOptions): Validation {
    return new MonitorAlarm(options);
  }

  /** @internal */
  readonly _label: string;

  /** @internal */
  protected constructor(options: ValidationBaseOptions) {
    this._label = options.label ?? 'Unlabelled';
  }

  /** @internal */
  abstract _bind(scope: cdk.Construct, id: string): ValidationConfig;
}

/** @internal */
class AlwaysSucceeds extends Validation {
  constructor() {
    super({ label: 'AlwaysSucceeds' });
  }

  _bind(scope: cdk.Construct, id: string): ValidationConfig {
    return {
      chainable: new sfn.Pass(scope, id, {
        result: sfn.Result.fromString('Winner'),
      }),
    };
  }
}

/** @internal */
class AlwaysFails extends Validation {
  constructor() {
    super({ label: 'AlwaysFails' });
  }

  _bind(scope: cdk.Construct, id: string): ValidationConfig {
    return {
      chainable: new sfn.Fail(scope, id),
    };
  }
}

/**
 * Base options for validations
 */
export interface ValidationBaseOptions {
  /**
   * Label references to this validation for easier identification.
   * @default - Automatic label based on the validation type
   */
  readonly label?: string;
}

/**
 * Base options for Fargate-based validations.
 */
export interface FargateValidationBaseOptions extends ValidationBaseOptions {
  /**
   * The cluster to create the task on.
   */
  readonly cluster: ecs.ICluster;

  /**
   * Container overrides.
   */
  readonly containerOverrides?: sfn_tasks.ContainerOverride[];

  /**
   * Select subnets in which tasks will run.
   */
  readonly subnets?: ec2.SubnetSelection;

  /**
   * Give tasks a security group.
   */
  readonly securityGroups?: ec2.ISecurityGroup[];

  /**
   * Assign tasks public IP addresses.
   */
  readonly assignPublicIp?: boolean;
}

/**
 * Options for Fargate task validations.
 */
export interface FargateTaskSucceedsOptions extends FargateValidationBaseOptions {
  readonly taskDefinition: ecs.TaskDefinition;
}

/** @internal */
class FargateTaskSucceeds extends Validation {
  constructor(private readonly options: FargateTaskSucceedsOptions) {
    super({
      label: options.label ?? 'FargateTaskSucceeds',
    });
  }

  _bind(scope: cdk.Construct, id: string): ValidationConfig {
    return {
      chainable: new sfn_tasks.EcsRunTask(scope, id, {
        launchTarget: new sfn_tasks.EcsFargateLaunchTarget(),
        ...this.options,
        integrationPattern: sfn.IntegrationPattern.RUN_JOB,
      }),
    };
  }
}

/**
 * Props for `FargateValidationFactory`
 */
export interface FargateValidationFactoryProps extends FargateValidationBaseOptions {
}

/**
 * Base options for running containers.
 */
export interface FargateValidationRunContainerOptions extends ValidationBaseOptions {
  /**
   * Container image to run.
   */
  readonly image: ecs.ContainerImage;

  /**
   * Run this container command.
   * @default - use the image's default.
   */
  readonly command?: string[];
}

/**
 * A convenience tool for creating Fargate-based validations.
 */
export class FargateValidationFactory extends cdk.Construct {
  private readonly options: FargateValidationFactoryProps;
  private readonly securityGroups: ec2.ISecurityGroup[];
  private taskDefinitionIndex = -1;

  constructor(scope: cdk.Construct, id: string, props: FargateValidationFactoryProps) {
    super(scope, id);
    this.options = props;
    // By default, no security group.
    this.securityGroups = props.securityGroups ?? [];
  }

  private obtainTaskDefinitionId() {
    this.taskDefinitionIndex += 1;
    return `TaskDefinition${this.taskDefinitionIndex}`;
  }

  /**
   * Produce a validation that uses a container image to run a command.
   */
  runContainer(options: FargateValidationRunContainerOptions): Validation {
    const taskDefinition = new ecs.FargateTaskDefinition(this, this.obtainTaskDefinitionId(), {
      cpu: 256,
      memoryLimitMiB: 512,
    });

    taskDefinition.addContainer('script', {
      image: options.image,
      command: options.command,
    });

    // Use the given label over the factory's default label.
    const label = options.label ?? this.options.label;

    return Validation.fargateTaskSucceeds({
      ...this.options,
      label: label,
      securityGroups: this.securityGroups,
      taskDefinition,
    });
  }
}

/**
 * Options for monitoring alarms.
 */
export interface MonitorAlarmOptions extends ValidationBaseOptions {
  /**
   * The alarm to monitor.
   */
  readonly alarm: cw.IAlarm;

  /**
   * The length of time to monitor the alarm.
   * @default - one minute
   */
  readonly duration?: cdk.Duration;
}

class MonitorAlarm extends Validation {
  private readonly alarm: cw.IAlarm;
  private readonly timeout: cdk.Duration;

  constructor(options: MonitorAlarmOptions) {
    super({
      label: options.label ?? 'MonitorAlarm',
    });

    this.alarm = options.alarm;
    this.timeout = options.duration ?? cdk.Duration.minutes(1);
  }

  _bind(scope: cdk.Construct, id: string): ValidationConfig {
    const privateScope = new cdk.Construct(scope, id);
    const alarmMonitor = new SingletonAlarmMonitor(privateScope, 'AlarmMonitor');

    const chainable = new sfn_tasks.StepFunctionsStartExecution(privateScope, id, {
      stateMachine: alarmMonitor.stateMachine,
      input: alarmMonitor.createTaskInput({
        alarm: this.alarm,
        duration: this.timeout,
      }),
      integrationPattern: sfn.IntegrationPattern.RUN_JOB,
    });

    return {
      chainable,
    };
  }
}

/**
 * Options for step function validations
 */
export interface StateMachineExecutionSucceedsOptions extends ValidationBaseOptions {
  /** The state machine to execute */
  readonly stateMachine: sfn.IStateMachine;

  /**
   * Input for the state machine's execution
   * @default - no input given
   */
  readonly input?: sfn.TaskInput;
}

class StateMachineExecutionSucceeds extends Validation {
  constructor(private readonly options: StateMachineExecutionSucceedsOptions) {
    super({
      label: options.label ?? 'StateMachineExecutionSucceeds',
    });
  }

  _bind(scope: cdk.Construct, id: string): ValidationConfig {
    const chainable = new sfn_tasks.StepFunctionsStartExecution(scope, id, {
      stateMachine: this.options.stateMachine,
      input: this.options.input,
      integrationPattern: sfn.IntegrationPattern.RUN_JOB,
    });

    return {
      chainable,
    };
  }
}