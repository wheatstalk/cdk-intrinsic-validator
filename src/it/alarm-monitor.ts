import * as cdk from '@aws-cdk/core';
import { AlarmMonitor } from '../alarm-monitor';
import { TestAlarms } from './test-alarms';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'AlarmMonitor', {
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  },
});

new AlarmMonitor(stack, 'AlarmMonitor');
new TestAlarms(stack, 'TestAlarms');

// Submit the JSON for the Test Alarms to the state machine to test this thing.