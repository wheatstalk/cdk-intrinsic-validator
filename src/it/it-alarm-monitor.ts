import * as cdk from '@aws-cdk/core';
import { AlarmMonitor } from '../alarm-monitor';
import { TestAlarms } from './test-alarms';

/** @internal */
export class ItAlarmMonitor extends cdk.Stack {
  constructor(scope: cdk.Construct, props: cdk.StackProps = {}) {
    super(scope, 'ItAlarmMonitor', props);

    new AlarmMonitor(this, 'AlarmMonitor');
    new TestAlarms(this, 'TestAlarms');

    // Submit the JSON for the Test Alarms to the state machine to test this thing.
  }
}

if (!module.parent) {
  const app = new cdk.App();
  new ItAlarmMonitor(app, {
    env: {
      account: process.env.CDK_DEFAULT_ACCOUNT,
      region: process.env.CDK_DEFAULT_REGION,
    },
  });
}
