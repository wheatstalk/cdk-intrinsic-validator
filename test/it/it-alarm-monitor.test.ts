import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { DisableRandomnessContextKey } from '../../src';
import { ItAlarmMonitor } from '../../src/it/it-alarm-monitor';

test('snapshot', () => {
  const app = new cdk.App({
    context: {
      [DisableRandomnessContextKey]: true,
    },
  });

  const stack = new ItAlarmMonitor(app);
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});