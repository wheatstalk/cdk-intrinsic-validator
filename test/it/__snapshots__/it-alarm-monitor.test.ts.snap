// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
Object {
  "Outputs": Object {
    "TestAlarmsAlwaysAlarmingExecution4B32A340": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "{\\"AlarmName\\":\\"",
            Object {
              "Ref": "TestAlarmsAlwaysAlarming69FED027",
            },
            "\\",\\"MonitoringSeconds\\":31}",
          ],
        ],
      },
    },
    "TestAlarmsNeverAlarmingExecutionF5D1776E": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "{\\"AlarmName\\":\\"",
            Object {
              "Ref": "TestAlarmsNeverAlarmingHopefullyFD878B05",
            },
            "\\",\\"MonitoringSeconds\\":60}",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters0832174fde72cd971a27479dcc997f3bbc8c5edcb918cc022ae4b956b6dc0adfArtifactHash18ED610C": Object {
      "Description": "Artifact hash for asset \\"0832174fde72cd971a27479dcc997f3bbc8c5edcb918cc022ae4b956b6dc0adf\\"",
      "Type": "String",
    },
    "AssetParameters0832174fde72cd971a27479dcc997f3bbc8c5edcb918cc022ae4b956b6dc0adfS3BucketCC6F037C": Object {
      "Description": "S3 bucket for asset \\"0832174fde72cd971a27479dcc997f3bbc8c5edcb918cc022ae4b956b6dc0adf\\"",
      "Type": "String",
    },
    "AssetParameters0832174fde72cd971a27479dcc997f3bbc8c5edcb918cc022ae4b956b6dc0adfS3VersionKeyA80BE83A": Object {
      "Description": "S3 key for asset version \\"0832174fde72cd971a27479dcc997f3bbc8c5edcb918cc022ae4b956b6dc0adf\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24ArtifactHashBA91B77F": Object {
      "Description": "Artifact hash for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5": Object {
      "Description": "S3 bucket for asset \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
    "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861": Object {
      "Description": "S3 key for asset version \\"67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "AlarmMonitorCheckAlarmStatusFunction38993418": Object {
      "DependsOn": Array [
        "AlarmMonitorCheckAlarmStatusFunctionServiceRoleDefaultPolicyF834009A",
        "AlarmMonitorCheckAlarmStatusFunctionServiceRole0FE14751",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0832174fde72cd971a27479dcc997f3bbc8c5edcb918cc022ae4b956b6dc0adfS3BucketCC6F037C",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0832174fde72cd971a27479dcc997f3bbc8c5edcb918cc022ae4b956b6dc0adfS3VersionKeyA80BE83A",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0832174fde72cd971a27479dcc997f3bbc8c5edcb918cc022ae4b956b6dc0adfS3VersionKeyA80BE83A",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "check-alarm-status.checkAlarmStatus",
        "Role": Object {
          "Fn::GetAtt": Array [
            "AlarmMonitorCheckAlarmStatusFunctionServiceRole0FE14751",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "AlarmMonitorCheckAlarmStatusFunctionLogRetention1E4E68E2": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "AlarmMonitorCheckAlarmStatusFunction38993418",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "AlarmMonitorCheckAlarmStatusFunctionServiceRole0FE14751": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AlarmMonitorCheckAlarmStatusFunctionServiceRoleDefaultPolicyF834009A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:DescribeAlarms",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AlarmMonitorCheckAlarmStatusFunctionServiceRoleDefaultPolicyF834009A",
        "Roles": Array [
          Object {
            "Ref": "AlarmMonitorCheckAlarmStatusFunctionServiceRole0FE14751",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "AlarmMonitorStateMachineC704D11C": Object {
      "DependsOn": Array [
        "AlarmMonitorStateMachineRoleDefaultPolicy7C3EF13E",
        "AlarmMonitorStateMachineRole19D08AAF",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"CheckAlarmStatus\\",\\"States\\":{\\"CheckAlarmStatus\\":{\\"Next\\":\\"AlarmStatusChoice\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2},{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"MaxAttempts\\":3}],\\"Catch\\":[{\\"ErrorEquals\\":[\\"States.ALL\\"],\\"Next\\":\\"BugState\\"}],\\"Type\\":\\"Task\\",\\"ResultPath\\":\\"$.CheckAlarmStatus\\",\\"ResultSelector\\":{\\"Status.$\\":\\"$.Payload.Status\\"},\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "AlarmMonitorCheckAlarmStatusFunction38993418",
                  "Arn",
                ],
              },
              "\\",\\"Payload\\":{\\"AlarmName.$\\":\\"$.AlarmName\\",\\"MonitoringDurationSeconds.$\\":\\"$.MonitoringDurationSeconds\\",\\"StartTime.$\\":\\"$$.Execution.StartTime\\"}}},\\"DelayThenNextCheckStatus\\":{\\"Type\\":\\"Wait\\",\\"Seconds\\":30,\\"Next\\":\\"CheckAlarmStatus\\"},\\"AlarmStatusChoice\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.CheckAlarmStatus.Status\\",\\"StringEquals\\":\\"ALARM\\",\\"Next\\":\\"FailBecauseAlarming\\"},{\\"Variable\\":\\"$.CheckAlarmStatus.Status\\",\\"StringEquals\\":\\"MONITORING\\",\\"Next\\":\\"DelayThenNextCheckStatus\\"},{\\"Variable\\":\\"$.CheckAlarmStatus.Status\\",\\"StringEquals\\":\\"NO_ALARM\\",\\"Next\\":\\"NoAlarm\\"}],\\"Default\\":\\"BugState\\"},\\"BugState\\":{\\"Type\\":\\"Fail\\"},\\"FailBecauseAlarming\\":{\\"Type\\":\\"Fail\\"},\\"NoAlarm\\":{\\"Type\\":\\"Succeed\\"}}}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "AlarmMonitorStateMachineRole19D08AAF",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "AlarmMonitorStateMachineRole19D08AAF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "AlarmMonitorStateMachineRoleDefaultPolicy7C3EF13E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "AlarmMonitorCheckAlarmStatusFunction38993418",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AlarmMonitorStateMachineRoleDefaultPolicy7C3EF13E",
        "Roles": Array [
          Object {
            "Ref": "AlarmMonitorStateMachineRole19D08AAF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3Bucket4D46ABB5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24S3VersionKeyB0F28861",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestAlarmsAlwaysAlarming69FED027": Object {
      "Properties": Object {
        "ComparisonOperator": "LessThanThreshold",
        "Dimensions": Array [
          Object {
            "Name": "Currency",
            "Value": "USD",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "EstimatedCharges",
        "Namespace": "AWS/Billing",
        "Period": 32400,
        "Statistic": "Maximum",
        "Threshold": 99999999999,
        "TreatMissingData": "breaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestAlarmsNeverAlarmingHopefullyFD878B05": Object {
      "Properties": Object {
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": Array [
          Object {
            "Name": "Currency",
            "Value": "USD",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "EstimatedCharges",
        "Namespace": "AWS/Billing",
        "Period": 32400,
        "Statistic": "Maximum",
        "Threshold": 99999999999,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;
